File: VB/VB/E968.Module/Updater.vb
Imports DatabaseUserSettings
Imports DevExpress.ExpressApp
Imports DevExpress.Data.Filtering
Imports DevExpress.ExpressApp.Security
Imports DevExpress.ExpressApp.Updating
Imports DatabaseUserSettings.BusinessObjects
Imports DevExpress.ExpressApp.Security.Strategy

Namespace E968.Module
    Public Class Updater
        Inherits ModuleUpdater

        Public Sub New(ByVal objectSpace As IObjectSpace, ByVal currentDBVersion As Version)
            MyBase.New(objectSpace, currentDBVersion)
        End Sub
        Public Overrides Sub UpdateDatabaseAfterUpdateSchema()
            MyBase.UpdateDatabaseAfterUpdateSchema()
            Dim adminRole As SecuritySystemRole = ObjectSpace.FindObject(Of SecuritySystemRole)(New BinaryOperator("Name", SecurityStrategy.AdministratorRoleName))
            If adminRole Is Nothing Then
                adminRole = ObjectSpace.CreateObject(Of SecuritySystemRole)()
                adminRole.Name = SecurityStrategy.AdministratorRoleName
                adminRole.IsAdministrative = True
                adminRole.Save()
            End If
            Dim adminUser As SecuritySystemUser = ObjectSpace.FindObject(Of SecuritySystemUser)(New BinaryOperator("UserName", "Sam"))
            If adminUser Is Nothing Then
                adminUser = ObjectSpace.CreateObject(Of SecuritySystemUser)()
                adminUser.UserName = "Sam"
                adminUser.SetPassword("")
                adminUser.Roles.Add(adminRole)
            End If
            Dim userJohn As SecuritySystemUser = ObjectSpace.FindObject(Of SecuritySystemUser)(New BinaryOperator("UserName", "John"))
            If userJohn Is Nothing Then
                userJohn = ObjectSpace.CreateObject(Of SecuritySystemUser)()
                userJohn.UserName = "John"
                userJohn.IsActive = True
                userJohn.Roles.Add(GetDefaultRole())
                userJohn.Save()
            End If
            Dim configuratorUser As SecuritySystemUser = ObjectSpace.FindObject(Of SecuritySystemUser)(New BinaryOperator("UserName", DatabaseUserSettingsModule.ConfiguratorUserName))
            If configuratorUser Is Nothing Then
                configuratorUser = ObjectSpace.CreateObject(Of SecuritySystemUser)()
                configuratorUser.UserName = DatabaseUserSettingsModule.ConfiguratorUserName
                configuratorUser.IsActive = True
                configuratorUser.Roles.Add(GetConfiguratorRole())
                configuratorUser.Roles.Add(adminRole)
                configuratorUser.Save()
            End If
            ObjectSpace.CommitChanges()
        End Sub
        Private Function GetDefaultRole() As SecuritySystemRole
            Dim defaultRole As SecuritySystemRole = ObjectSpace.FindObject(Of SecuritySystemRole)(New BinaryOperator("Name", "Default"))
            If defaultRole Is Nothing Then
                defaultRole = ObjectSpace.CreateObject(Of SecuritySystemRole)()
                defaultRole.Name = "Default"

                Dim userPermissions As SecuritySystemTypePermissionObject = ObjectSpace.CreateObject(Of SecuritySystemTypePermissionObject)()
                userPermissions.TargetType = GetType(SecuritySystemUser)
                userPermissions.AllowNavigate = True
                defaultRole.TypePermissions.Add(userPermissions)
                Dim myDetailsPermission As SecuritySystemObjectPermissionsObject = ObjectSpace.CreateObject(Of SecuritySystemObjectPermissionsObject)()
                myDetailsPermission.Criteria = "[Oid] = CurrentUserId()"
                myDetailsPermission.AllowNavigate = True
                myDetailsPermission.AllowRead = True
                userPermissions.ObjectPermissions.Add(myDetailsPermission)
                Dim ownPasswordPermission As SecuritySystemMemberPermissionsObject = ObjectSpace.CreateObject(Of SecuritySystemMemberPermissionsObject)()
                ownPasswordPermission.Members = "ChangePasswordOnFirstLogon; StoredPassword"
                ownPasswordPermission.AllowWrite = True
                userPermissions.MemberPermissions.Add(ownPasswordPermission)

                Dim rolePermissions As SecuritySystemTypePermissionObject = ObjectSpace.CreateObject(Of SecuritySystemTypePermissionObject)()
                rolePermissions.TargetType = GetType(SecuritySystemRole)
                rolePermissions.AllowNavigate = True
                defaultRole.TypePermissions.Add(rolePermissions)

                Dim defaultRolePermission As SecuritySystemObjectPermissionsObject = ObjectSpace.CreateObject(Of SecuritySystemObjectPermissionsObject)()
                defaultRolePermission.Criteria = "[Name] = 'Default'"
                defaultRolePermission.AllowRead = True
                rolePermissions.ObjectPermissions.Add(defaultRolePermission)

                defaultRole.TypePermissions.Add(DatabaseUserSettings.DatabaseUserSettingsModule.CreateUserSettingsAspectPermissions(ObjectSpace))
                defaultRole.TypePermissions.Add(DatabaseUserSettings.DatabaseUserSettingsModule.CreateUserSettingsPermissions(ObjectSpace))
                defaultRole.Save()
            End If
            Return defaultRole
        End Function
        Private Function GetConfiguratorRole() As SecuritySystemRole
            Dim configuratorRole As SecuritySystemRole = ObjectSpace.FindObject(Of SecuritySystemRole)(New BinaryOperator("Name", DatabaseUserSettingsModule.ConfiguratorRoleName))
            If configuratorRole Is Nothing Then
                configuratorRole = ObjectSpace.CreateObject(Of SecuritySystemRole)()
                configuratorRole.Name = DatabaseUserSettingsModule.ConfiguratorRoleName
                configuratorRole.ChildRoles.Add(GetDefaultRole())

                Dim userSettingsAspectPermissions As SecuritySystemTypePermissionObject = ObjectSpace.CreateObject(Of SecuritySystemTypePermissionObject)()
                userSettingsAspectPermissions.TargetType = GetType(XPUserSettingsAspect)
                userSettingsAspectPermissions.AllowCreate = True
                userSettingsAspectPermissions.AllowDelete = True
                userSettingsAspectPermissions.AllowNavigate = True
                userSettingsAspectPermissions.AllowRead = True
                userSettingsAspectPermissions.AllowWrite = True
                configuratorRole.TypePermissions.Add(userSettingsAspectPermissions)

                Dim userSettingsPermissions As SecuritySystemTypePermissionObject = ObjectSpace.CreateObject(Of SecuritySystemTypePermissionObject)()
                userSettingsPermissions.TargetType = GetType(XPUserSettings)
                userSettingsPermissions.AllowCreate = True
                userSettingsPermissions.AllowDelete = True
                userSettingsPermissions.AllowNavigate = True
                userSettingsPermissions.AllowRead = True
                userSettingsPermissions.AllowWrite = True
                configuratorRole.TypePermissions.Add(userSettingsPermissions)

                Dim userPermissions As SecuritySystemTypePermissionObject = ObjectSpace.CreateObject(Of SecuritySystemTypePermissionObject)()
                userPermissions.TargetType = GetType(SecuritySystemUser)
                configuratorRole.TypePermissions.Add(userPermissions)
                Dim userNamePermission As SecuritySystemMemberPermissionsObject = ObjectSpace.CreateObject(Of SecuritySystemMemberPermissionsObject)()
                userNamePermission.Members = "UserName"
                userNamePermission.AllowRead = True
                userPermissions.MemberPermissions.Add(userNamePermission)
                configuratorRole.Save()
            End If
            Return configuratorRole
        End Function
    End Class
End Namespace

File: CS/CS/E968.Module/Updater.cs
using System;
using DatabaseUserSettings;
using DevExpress.ExpressApp;
using DevExpress.Data.Filtering;
using DevExpress.ExpressApp.Security;
using DevExpress.ExpressApp.Updating;
using DatabaseUserSettings.BusinessObjects;
using DevExpress.ExpressApp.Security.Strategy;

namespace E968.Module {
    public class Updater : ModuleUpdater {
        public Updater(IObjectSpace objectSpace, Version currentDBVersion) : base(objectSpace, currentDBVersion) { }
        public override void UpdateDatabaseAfterUpdateSchema() {
            base.UpdateDatabaseAfterUpdateSchema();
            SecuritySystemRole adminRole = ObjectSpace.FindObject<SecuritySystemRole>(new BinaryOperator("Name", SecurityStrategy.AdministratorRoleName));
            if (adminRole == null) {
                adminRole = ObjectSpace.CreateObject<SecuritySystemRole>();
                adminRole.Name = SecurityStrategy.AdministratorRoleName;
                adminRole.IsAdministrative = true;
                adminRole.Save();
            }
            SecuritySystemUser adminUser = ObjectSpace.FindObject<SecuritySystemUser>(new BinaryOperator("UserName", "Sam"));
            if (adminUser == null) {
                adminUser = ObjectSpace.CreateObject<SecuritySystemUser>();
                adminUser.UserName = "Sam";
                adminUser.SetPassword("");
                adminUser.Roles.Add(adminRole);
            }
            SecuritySystemUser userJohn = ObjectSpace.FindObject<SecuritySystemUser>(new BinaryOperator("UserName", "John"));
            if (userJohn == null) {
                userJohn = ObjectSpace.CreateObject<SecuritySystemUser>();
                userJohn.UserName = "John";
                userJohn.IsActive = true;
                userJohn.Roles.Add(GetDefaultRole());
                userJohn.Save();
            }
            SecuritySystemUser configuratorUser = ObjectSpace.FindObject<SecuritySystemUser>(new BinaryOperator("UserName", DatabaseUserSettingsModule.ConfiguratorUserName));
            if (configuratorUser == null) {
                configuratorUser = ObjectSpace.CreateObject<SecuritySystemUser>();
                configuratorUser.UserName = DatabaseUserSettingsModule.ConfiguratorUserName;
                configuratorUser.IsActive = true;
                configuratorUser.Roles.Add(GetConfiguratorRole());
                configuratorUser.Roles.Add(adminRole);
                configuratorUser.Save();
            }
            ObjectSpace.CommitChanges();
        }
        private SecuritySystemRole GetDefaultRole() {
            SecuritySystemRole defaultRole = ObjectSpace.FindObject<SecuritySystemRole>(new BinaryOperator("Name", "Default"));
            if (defaultRole == null) {
                defaultRole = ObjectSpace.CreateObject<SecuritySystemRole>();
                defaultRole.Name = "Default";

                SecuritySystemTypePermissionObject userPermissions = ObjectSpace.CreateObject<SecuritySystemTypePermissionObject>();
                userPermissions.TargetType = typeof(SecuritySystemUser);
                userPermissions.AllowNavigate = true;
                defaultRole.TypePermissions.Add(userPermissions);
                SecuritySystemObjectPermissionsObject myDetailsPermission = ObjectSpace.CreateObject<SecuritySystemObjectPermissionsObject>();
                myDetailsPermission.Criteria = "[Oid] = CurrentUserId()";
                myDetailsPermission.AllowNavigate = true;
                myDetailsPermission.AllowRead = true;
                userPermissions.ObjectPermissions.Add(myDetailsPermission);
                SecuritySystemMemberPermissionsObject ownPasswordPermission = ObjectSpace.CreateObject<SecuritySystemMemberPermissionsObject>();
                ownPasswordPermission.Members = "ChangePasswordOnFirstLogon; StoredPassword";
                ownPasswordPermission.AllowWrite = true;
                userPermissions.MemberPermissions.Add(ownPasswordPermission);

                SecuritySystemTypePermissionObject rolePermissions = ObjectSpace.CreateObject<SecuritySystemTypePermissionObject>();
                rolePermissions.TargetType = typeof(SecuritySystemRole);
                rolePermissions.AllowNavigate = true;
                defaultRole.TypePermissions.Add(rolePermissions);

                SecuritySystemObjectPermissionsObject defaultRolePermission = ObjectSpace.CreateObject<SecuritySystemObjectPermissionsObject>();
                defaultRolePermission.Criteria = "[Name] = 'Default'";
                defaultRolePermission.AllowRead = true;
                rolePermissions.ObjectPermissions.Add(defaultRolePermission);

                defaultRole.TypePermissions.Add(DatabaseUserSettings.DatabaseUserSettingsModule.CreateUserSettingsAspectPermissions(ObjectSpace));
                defaultRole.TypePermissions.Add(DatabaseUserSettings.DatabaseUserSettingsModule.CreateUserSettingsPermissions(ObjectSpace));
                defaultRole.Save();
            }
            return defaultRole;
        }
        private SecuritySystemRole GetConfiguratorRole() {
            SecuritySystemRole configuratorRole = ObjectSpace.FindObject<SecuritySystemRole>(new BinaryOperator("Name", DatabaseUserSettingsModule.ConfiguratorRoleName));
            if (configuratorRole == null) {
                configuratorRole = ObjectSpace.CreateObject<SecuritySystemRole>();
                configuratorRole.Name = DatabaseUserSettingsModule.ConfiguratorRoleName;
                configuratorRole.ChildRoles.Add(GetDefaultRole());

                SecuritySystemTypePermissionObject userSettingsAspectPermissions = ObjectSpace.CreateObject<SecuritySystemTypePermissionObject>();
                userSettingsAspectPermissions.TargetType = typeof(XPUserSettingsAspect);
                userSettingsAspectPermissions.AllowCreate = true;
                userSettingsAspectPermissions.AllowDelete = true;
                userSettingsAspectPermissions.AllowNavigate = true;
                userSettingsAspectPermissions.AllowRead = true;
                userSettingsAspectPermissions.AllowWrite = true;
                configuratorRole.TypePermissions.Add(userSettingsAspectPermissions);

                SecuritySystemTypePermissionObject userSettingsPermissions = ObjectSpace.CreateObject<SecuritySystemTypePermissionObject>();
                userSettingsPermissions.TargetType = typeof(XPUserSettings);
                userSettingsPermissions.AllowCreate = true;
                userSettingsPermissions.AllowDelete = true;
                userSettingsPermissions.AllowNavigate = true;
                userSettingsPermissions.AllowRead = true;
                userSettingsPermissions.AllowWrite = true;
                configuratorRole.TypePermissions.Add(userSettingsPermissions);

                SecuritySystemTypePermissionObject userPermissions = ObjectSpace.CreateObject<SecuritySystemTypePermissionObject>();
                userPermissions.TargetType = typeof(SecuritySystemUser);
                configuratorRole.TypePermissions.Add(userPermissions);
                SecuritySystemMemberPermissionsObject userNamePermission = ObjectSpace.CreateObject<SecuritySystemMemberPermissionsObject>();
                userNamePermission.Members = "UserName";
                userNamePermission.AllowRead = true;
                userPermissions.MemberPermissions.Add(userNamePermission);
                configuratorRole.Save();
            }
            return configuratorRole;
        }
    }
}